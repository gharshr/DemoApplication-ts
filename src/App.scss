.App {
  text-align: center;
}

.opaque {
  box-shadow: 0px 0px 10px;
  display: inline-block;
}

.like-text {
  border: hidden;
  display: inline-block;
}