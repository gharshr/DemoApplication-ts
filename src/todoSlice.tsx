import { createSlice } from  '@reduxjs/toolkit';

export interface todoNodeStructure {
  title : string,
  completed : boolean 
}

const { reducer, actions } = createSlice({
  name : 'todos',
  initialState : [{title : 'Checkout the Storage', completed : false}, {title : 'Check the passwords & Patternise', completed : true}] as todoNodeStructure[],
  reducers : {
    ADD_TODO : state => state,
    DELETE_TODO : state => state
  }
})

export default reducer;
export const { ADD_TODO, DELETE_TODO } = actions; 