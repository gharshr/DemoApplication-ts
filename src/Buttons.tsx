import React from 'react';
import { Col, Row, Button } from 'reactstrap';
import { ACTIVE_ALL,COMPLETE_ALL,DELETE_ALL } from './todoSlice';
import { connect } from 'react-redux'

export const Buttons = React.memo(function Buttons(props : any) {
  function onActiveAll() {
    fetch('/activeAll',{
      method : 'get'
    }).then(res => res.json()).then(data => {
      if(data.successfullOrNot) 
        props.ACTIVE_ALL()
      else
        alert(data.message)
    }).catch(err => console.log(err))
  }

  function onCompleteAll() {
    fetch('/completeAll',{
      method : 'get'
    }).then(res => res.json()).then(data => {
      if(data.successfullOrNot) 
        props.COMPLETE_ALL()
      else
        alert(data.message)
    }).catch(err => console.log(err))
  }

  function onDeleteAll() {
    fetch(`/todo/all`,{
      method : 'delete'
    }).then(res => res.json()).then(data => {
      if(data.successfullOrNot) 
        props.DELETE_ALL()
      else
        alert(data.message)
    }).catch(err => console.log(err))
  }

  console.log(props)
  return (
    <Row>
      <Col xs="4">
        <Row>
          <Col md={{offset : 3}}>
            <Button color="primary" size="sm" onClick={onActiveAll}>Add All</Button>
          </Col>
          <Col md={{offset : 1}}>
            <Button color="success" size="sm" onClick={onCompleteAll}>Complete All</Button>
          </Col>
          <Col md={{offset : 1}}>
            <Button color="danger" size="sm" onClick={onDeleteAll}>Delete All</Button>
          </Col>
        </Row>
      </Col>
    </Row>
  )
})

const mapDispatchToProps = { ACTIVE_ALL, COMPLETE_ALL, DELETE_ALL }

export default connect(null,mapDispatchToProps)(Buttons)