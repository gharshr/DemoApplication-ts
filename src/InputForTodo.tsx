import React, { useState, useEffect } from 'react'
import { Col, InputGroup, Input, InputGroupAddon, Button, Alert, Row } from 'reactstrap';
import { connect } from 'react-redux';
import { ADD_TODO, UPDATE_TODO, todoNodeStructure} from './todoSlice';
import { RootState } from '.';

const InputForTodo = React.memo(function (props : any = {editedTitle : ''}) {
    console.log(props.todos.find((todo : todoNodeStructure) => todo.id === props.id)?.title === props.editedTitle)
    const [toDoTitle, updateToDoTitle] = useState(props.editedTitle ? props.editedTitle : '');
    const [alertToDo, toggleToDoAlertVisibility] = useState(false);

    useEffect(() => {
        updateToDoTitle(props.editedTitle ? props.editedTitle : '')
    },[props.editedTitle])

    return (
        <Row>
            <Col xs="4">
                <InputGroup>
                    <Input type="text" disabled={props.type === "EDIT" ? true : false} id={props.type === "ADD" ? "add_todo_input" : "edit_todo_input"} placeholder={props.type === "ADD" ? "What do you want to do?" : "" } value={ toDoTitle } onChange={(e) => { updateToDoTitle(e.target.value); e.target.value.length < 5 ? toggleToDoAlertVisibility(true) : toggleToDoAlertVisibility(false)}}/>
                    <InputGroupAddon addonType="append">{props.type === "ADD" ? <Button color="primary" size="sm" id="todo_add_button" disabled={ toDoTitle.length < 5 ? true : false} onClick={() => { props.ADD_TODO({id: props.todos.length + 1, title : toDoTitle, completed : false}); updateToDoTitle('') }}>Add</Button>: <Button color="primary" size="sm" id="edit_todo_button" disabled={toDoTitle.length < 5 || props.todos.find((todo : todoNodeStructure) => todo.id === props.id)?.title === toDoTitle ? true : false} onClick={() => {props.UPDATE_TODO({id: props.id, title : toDoTitle}); (document.getElementById('edit_todo_input') as HTMLInputElement).disabled = true; (document.getElementById('edit_todo_button') as HTMLButtonElement).disabled = true; props.updateSelectedTitle(''); props.updateId(0)}}>Save</Button>}</InputGroupAddon>
                </InputGroup>
            </Col>
            <Col>
                <Alert isOpen={alertToDo} color="info">Length of the field should be greater than 5</Alert>
            </Col>
        </Row>
    )
})

const mapStateToProps = (state : RootState) => ({
    lastTodoId : state.lastTodoId
})

const mapDispatchToProps = { ADD_TODO, UPDATE_TODO }

export default connect(mapStateToProps,mapDispatchToProps)(InputForTodo);