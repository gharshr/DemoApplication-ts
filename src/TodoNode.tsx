import React from 'react';
import { connect } from 'react-redux'
import { DELETE_TODO, MARK_AS_COMPLETE } from './todoSlice';
import { Button, Input, Col, Row } from 'reactstrap'

// interface props extends todoNodeStructure {
//     updateToDoTitle : any,
//     updateId : any,
// }

const TodoNode = React.memo(function TodoNode(props : any ) {
    console.log(props)
    // const [completed,updateStatus] = useState(props.completed);
    return (
        <Row xs="4">
            <Col xs="1">
                <Input addon type="checkbox" checked={props.completed} onChange={() => props.MARK_AS_COMPLETE({id : props.id})}/>
            </Col>
            <Col xs="7"><b>{props.title}</b></Col>
            <Col md={{offset : 1}} xs="1">
                <Button color="warning" size="sm" onClick={() => { props.updateToDoTitle(props.title); props.updateId(props.id); (document.getElementById('edit_todo_input') as HTMLInputElement).disabled = false } }>Edit</Button>
            </Col>
            <Col xs="1">    
                <Button color="danger" size="sm" onClick={() => { props.DELETE_TODO({ id : props.id }) }}>Delete</Button>
            </Col>
        </Row>
    )
})

const mapDispatchToProps = { DELETE_TODO, MARK_AS_COMPLETE }

export default connect(null,mapDispatchToProps)(TodoNode);