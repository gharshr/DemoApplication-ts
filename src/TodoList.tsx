import React, { useState } from 'react'
import { ListGroup, ListGroupItem, Row, Col } from 'reactstrap';
import TodoNode from './TodoNode';
import { todoNodeStructure } from './todoSlice';
import InputForTodo from './InputForTodo';

const TodoList = React.memo(function TodoList(props : {todos : todoNodeStructure[]}) {
    const [editedTitle, updateSelectedTitle] = useState('');
    const [id,updateId] = useState(0);

    return (
        <ListGroup>
            {
                props.todos.map((todo : todoNodeStructure) => {
                    return (
                        <Row key={todo.id}>
                            <Col xs="4">
                                <ListGroupItem>
                                    <TodoNode title={todo.title} completed={todo.completed} id={todo.id} updateToDoTitle={updateSelectedTitle} updateId={updateId}/>
                                </ListGroupItem>
                            </Col>
                        </Row>
                    )
                })
            }
            <InputForTodo todos={props.todos} type="EDIT" id={id} editedTitle={editedTitle} updateSelectedTitle={updateSelectedTitle} updateId={updateId}></InputForTodo>
        </ListGroup>
    )
})

export default TodoList