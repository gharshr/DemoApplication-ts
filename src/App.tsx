import React, { useEffect } from 'react';
import TodoList from './TodoList'
import { RootState } from './index'
import { connect } from 'react-redux'
import Buttons from './Buttons'
import './App.scss';
import { Container, Badge } from 'reactstrap'
import InputForTodo from './InputForTodo';
import { todoNodeStructure, INITIALIZE_TODOS } from './todoSlice';

export function App(props : any = {INITIALIZE_TODOS : (a : RootState) => {}}) {
  
  useEffect(() => {
    console.log('called only once');
    fetch('/todos',{
      'method' : 'get',
    }).then(response => response.json()).then(data => { 
      console.log(data)
      props.INITIALIZE_TODOS({ todos : data, lastTodoId : data.length > 0 ? data.reduce((max : todoNodeStructure, node : todoNodeStructure) => {
        console.log(max)
        if(node.id > max.id)
          max = node
        return max
      }).id : 0})
    })
  },[])
  
  return (
    // <div>
    //   <h1>Hello People<p id="nine">This is just a demo test</p></h1>
    // </div>
    <Container fluid>
      <h1>todos - <Badge color="info">{props.todos.length}</Badge></h1>
      <InputForTodo todos={props.todos} type="ADD"></InputForTodo>
      <TodoList todos={props.todos}/>
      <Buttons/>
    </Container>
  );
}

const mapStateToProps = (state : RootState) => ({
  todos : state.todos
})

const mapDispatchToProps = { INITIALIZE_TODOS }

export default connect(mapStateToProps,mapDispatchToProps)(App);