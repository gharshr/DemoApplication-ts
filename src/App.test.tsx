import React from 'react';
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'
import { App } from './App';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import InputForTodo from './InputForTodo';
import TodoList from './TodoList';
import TodoNode from './TodoNode';
import { Provider } from 'react-redux';
import { ACTIVE_ALL, COMPLETE_ALL, DELETE_ALL } from './todoSlice'
import { configureStore } from '@reduxjs/toolkit';
import reducer from './todoSlice'
import { Buttons } from './Buttons';
import { Container } from 'reactstrap';

const store = configureStore({
  reducer
})

const Foo = (props : { width : number, height : number} ) => (
  <div>
    <input name="width" value={props.width} />
    <input name="height" value={props.height} />
  </div>
);

configure({ adapter: new Adapter() });

describe('React component test with Enzyme', () => {
  it('App renders without crashing', () => {
    const wrapper = shallow(<App todos={[]}/>);
    console.log(wrapper.find('h1').type())
  });

  // it('InputForTodo renders without crashing', () => {
  //   const wrapper = shallow(<Provider store={store}><InputForTodo/></Provider>);
  //   console.log(wrapper.debug());
  // });

  it('TodoList renders without crashing', () => {
    const wrapper = shallow(<TodoList todos={[]}/>);
    console.log(wrapper.debug());
  });

  // it('TodoNode renders without crashing', () => {
  //   const wrapper = shallow(<Provider store={store}><TodoNode/></Provider>);
  //   console.log(wrapper.debug());
  // });

  it('Buttons renders without crashing', () => {
    const wrapper = shallow(<Buttons ACTIVE_ALL={() => store.dispatch(ACTIVE_ALL())} COMPLETE_ALL={() => store.dispatch(COMPLETE_ALL())} DELETE_ALL={() => store.dispatch(DELETE_ALL())}/>);
    console.log(wrapper.find('Button').forEach((node) => {
      node.simulate('click');
      store.subscribe(() => {
        console.log(store.getState().todos)
        switch(node.children().text()) {
          case 'Add All' : expect(store.getState().todos.every(todo => todo.completed === false)).toEqual(true);
          case 'Complete All' : expect(store.getState().todos.every(todo => todo.completed === true)).toEqual(true);
          case 'Delete All' : expect(store.getState().todos.length === 0);
        }
      })
    }))
  });
});

// describe('React Components test with Jest', () => {
//   test('App Component with Snapshot', () => {
//     const component = renderer.create(
//       <Provider store={store}><App todos={[]}/></Provider>
//     );
  
//     let tree = component.toJSON();
//     expect(tree).toMatchSnapshot();
//   })
  
//   test('InputForTodo Component with Snapshot', () => {
//     const component = renderer.create(
//       <Provider store={store}><InputForTodo todos={[]} /></Provider>
//     );
  
//     let tree = component.toJSON();
//     expect(tree).toMatchSnapshot();
//   })
  
//   test('TodoList Component with Snapshot', () => {
//     const component = renderer.create(
//       <Provider store={store}><TodoList todos={[]} /></Provider>
//     );
  
//     let tree = component.toJSON();
//     expect(tree).toMatchSnapshot();
//   })
  
//   test('TodoNode Component with Snapshot', () => {
//     const component = renderer.create(
//       <Provider store={store}><TodoNode /></Provider>
//     );
  
//     let tree = component.toJSON();
//     expect(tree).toMatchSnapshot();
//   })
  
//   test('Buttons Component with Snapshot', () => {
//     const component = renderer.create(
//       <Provider store={store}><Buttons /></Provider>
//     );
  
//     let tree = component.toJSON();
//     expect(tree).toMatchSnapshot();
//   })
// })