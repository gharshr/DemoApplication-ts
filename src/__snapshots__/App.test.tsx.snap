// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React Components test with Jest App Component with Snapshot 1`] = `
<div
  className="container-fluid"
>
  <h1>
    todos - 
    <span
      className="badge badge-info"
    >
      0
    </span>
  </h1>
  <div
    className="row"
  >
    <div
      className="col-4"
    >
      <div
        className="rbt"
        style={
          Object {
            "outline": "none",
            "position": "relative",
          }
        }
        tabIndex={-1}
      >
        <div
          style={
            Object {
              "display": "flex",
              "flex": 1,
              "height": "100%",
              "position": "relative",
            }
          }
        >
          <input
            aria-autocomplete="both"
            aria-expanded={false}
            aria-haspopup="listbox"
            autoComplete="off"
            className="rbt-input-main form-control rbt-input"
            disabled={false}
            id="add_todo_input"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="What do you want to do?"
            role="combobox"
            type="text"
            value=""
          />
          <input
            aria-hidden={true}
            className="rbt-input-hint"
            readOnly={true}
            style={
              Object {
                "backgroundColor": "transparent",
                "borderColor": "transparent",
                "boxShadow": "none",
                "color": "rgba(0, 0, 0, 0.35)",
                "left": 0,
                "pointerEvents": "none",
                "position": "absolute",
                "top": 0,
                "width": "100%",
              }
            }
            tabIndex={-1}
            value=""
          />
        </div>
      </div>
      <div
        className="input-group"
      />
    </div>
    <div
      className="col"
    />
  </div>
  <ul
    className="list-group"
  >
    <div
      className="row"
    >
      <div
        className="col-4"
      >
        <div
          className="rbt"
          style={
            Object {
              "outline": "none",
              "position": "relative",
            }
          }
          tabIndex={-1}
        >
          <div
            style={
              Object {
                "display": "flex",
                "flex": 1,
                "height": "100%",
                "position": "relative",
              }
            }
          >
            <input
              aria-autocomplete="both"
              aria-expanded={false}
              aria-haspopup="listbox"
              autoComplete="off"
              className="rbt-input-main form-control rbt-input"
              disabled={true}
              id="edit_todo_input"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder=""
              role="combobox"
              type="text"
              value=""
            />
            <input
              aria-hidden={true}
              className="rbt-input-hint"
              readOnly={true}
              style={
                Object {
                  "backgroundColor": "transparent",
                  "borderColor": "transparent",
                  "boxShadow": "none",
                  "color": "rgba(0, 0, 0, 0.35)",
                  "left": 0,
                  "pointerEvents": "none",
                  "position": "absolute",
                  "top": 0,
                  "width": "100%",
                }
              }
              tabIndex={-1}
              value=""
            />
          </div>
        </div>
        <div
          className="input-group"
        />
      </div>
      <div
        className="col"
      />
    </div>
  </ul>
  <div
    className="row"
  >
    <div
      className="col-4"
    >
      <div
        className="row"
      >
        <div
          className="offset-md-3"
        >
          <button
            aria-label={null}
            className="btn btn-primary btn-sm"
            onClick={[Function]}
            type="button"
          >
            Add All
          </button>
        </div>
        <div
          className="offset-md-1"
        >
          <button
            aria-label={null}
            className="btn btn-success btn-sm"
            onClick={[Function]}
            type="button"
          >
            Complete All
          </button>
        </div>
        <div
          className="offset-md-1"
        >
          <button
            aria-label={null}
            className="btn btn-danger btn-sm"
            onClick={[Function]}
            type="button"
          >
            Delete All
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`React Components test with Jest Buttons Component with Snapshot 1`] = `
<div
  className="row"
>
  <div
    className="col-4"
  >
    <div
      className="row"
    >
      <div
        className="offset-md-3"
      >
        <button
          aria-label={null}
          className="btn btn-primary btn-sm"
          onClick={[Function]}
          type="button"
        >
          Add All
        </button>
      </div>
      <div
        className="offset-md-1"
      >
        <button
          aria-label={null}
          className="btn btn-success btn-sm"
          onClick={[Function]}
          type="button"
        >
          Complete All
        </button>
      </div>
      <div
        className="offset-md-1"
      >
        <button
          aria-label={null}
          className="btn btn-danger btn-sm"
          onClick={[Function]}
          type="button"
        >
          Delete All
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`React Components test with Jest InputForTodo Component with Snapshot 1`] = `
<div
  className="row"
>
  <div
    className="col-4"
  >
    <div
      className="rbt"
      style={
        Object {
          "outline": "none",
          "position": "relative",
        }
      }
      tabIndex={-1}
    >
      <div
        style={
          Object {
            "display": "flex",
            "flex": 1,
            "height": "100%",
            "position": "relative",
          }
        }
      >
        <input
          aria-autocomplete="both"
          aria-expanded={false}
          aria-haspopup="listbox"
          autoComplete="off"
          className="rbt-input-main form-control rbt-input"
          disabled={false}
          id="add_todo_input"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder=""
          role="combobox"
          type="text"
          value=""
        />
        <input
          aria-hidden={true}
          className="rbt-input-hint"
          readOnly={true}
          style={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "boxShadow": "none",
              "color": "rgba(0, 0, 0, 0.35)",
              "left": 0,
              "pointerEvents": "none",
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
          tabIndex={-1}
          value=""
        />
      </div>
    </div>
    <div
      className="input-group"
    />
  </div>
  <div
    className="col"
  />
</div>
`;

exports[`React Components test with Jest TodoList Component with Snapshot 1`] = `
<ul
  className="list-group"
>
  <div
    className="row"
  >
    <div
      className="col-4"
    >
      <div
        className="rbt"
        style={
          Object {
            "outline": "none",
            "position": "relative",
          }
        }
        tabIndex={-1}
      >
        <div
          style={
            Object {
              "display": "flex",
              "flex": 1,
              "height": "100%",
              "position": "relative",
            }
          }
        >
          <input
            aria-autocomplete="both"
            aria-expanded={false}
            aria-haspopup="listbox"
            autoComplete="off"
            className="rbt-input-main form-control rbt-input"
            disabled={true}
            id="edit_todo_input"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder=""
            role="combobox"
            type="text"
            value=""
          />
          <input
            aria-hidden={true}
            className="rbt-input-hint"
            readOnly={true}
            style={
              Object {
                "backgroundColor": "transparent",
                "borderColor": "transparent",
                "boxShadow": "none",
                "color": "rgba(0, 0, 0, 0.35)",
                "left": 0,
                "pointerEvents": "none",
                "position": "absolute",
                "top": 0,
                "width": "100%",
              }
            }
            tabIndex={-1}
            value=""
          />
        </div>
      </div>
      <div
        className="input-group"
      />
    </div>
    <div
      className="col"
    />
  </div>
</ul>
`;

exports[`React Components test with Jest TodoNode Component with Snapshot 1`] = `
<div
  className="row row-cols-4"
>
  <div
    className="col-1"
  >
    <input
      className=""
      onChange={[Function]}
      type="checkbox"
    />
  </div>
  <div
    className="col-7"
  >
    <b />
  </div>
  <div
    className="col-1 offset-md-1"
  >
    <button
      aria-label={null}
      className="btn btn-warning btn-sm"
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </div>
  <div
    className="col-1"
  >
    <button
      aria-label={null}
      className="btn btn-danger btn-sm"
      onClick={[Function]}
      type="button"
    >
      Delete
    </button>
  </div>
</div>
`;
